tiledb/sm/query/query.cc

```diff
@@ -612,779 +610,47 @@ tiledb::sm::Status Query::from_capnp(rest::capnp::Query::Reader* query) {
-                    std::to_string(buffer_length) + ")");
-              }
-              uint8_t* existing_buffer_casted =
-                  static_cast<uint8_t*>(existingBuffer);
-              for (uint i = 0; i < buffer_length; i++) {
-                existing_buffer_casted[i] = newBuffer[i];
-              }
-
-              if (buffer.hasBufferOffset()) {
-                auto buffer_offset = buffer.getBufferOffset();
-                if (*existingBufferOffsetSize / sizeof(uint64_t) !=
-                    buffer_offset.size()) {
-                  return Status::QueryError(
-                      "Existing buffer_var_ in query object is different size "
-                      "(" +
-                      std::to_string(*existingBufferOffsetSize) +
-                      ") vs new query object buffer_var size (" +
-                      std::to_string(buffer_offset.size()) + ")");
-                }
-                uint64_t* existing_buffer_var_caster =
-                    static_cast<uint64_t*>(existingBufferOffset);
-                for (uint i = 0; i < buffer_offset.size(); i++) {
-                  existing_buffer_var_caster[i] = buffer_offset[i];
-                }
-              }
-
-            } else {
-              // Get buffer
-              std::vector<uint8_t>* new_buffer =
-                  new std::vector<uint8_t>(buffer_length);
-              for (uint i = 0; i < buffer_length; i++) {
-                (*new_buffer)[i] = newBuffer[i];
-              }
-              uint64_t* buffer_size = new uint64_t(buffer_length * type_size);
-
-              // Check for offset buffer
-              if (buffer.hasBufferOffset()) {
-                auto buffer_offset = buffer.getBufferOffset();
-                std::vector<uint64_t>* new_buffer_offset =
-                    new std::vector<uint64_t>(buffer_offset.size());
-                for (uint i = 0; i < buffer_offset.size(); i++) {
-                  (*new_buffer_offset)[i] = buffer_offset[i];
-                }
-
-                uint64_t* new_buffer_offset_size =
-                    new uint64_t(new_buffer_offset->size() * sizeof(uint64_t));
-                set_buffer(
-                    attr->name(),
-                    new_buffer_offset->data(),
-                    new_buffer_offset_size,
-                    new_buffer->data(),
-                    buffer_size);
-              } else
-                set_buffer(attr->name(), new_buffer->data(), buffer_size);
-            }
+        if (offset_buffer_size_in_bytes) {
+          if (*existingBufferOffsetSize >= offset_buffer_size_in_bytes) {
+            return Status::QueryError(
+                "Existing buffer_var_ in query object is smaller size "
+                "(" +
+                std::to_string(*existingBufferOffsetSize) +
+                ") vs new query object buffer_var size (" +
+                std::to_string(offset_buffer_size_in_bytes) + ")");
           }
-          break;
+          memcpy(
```

Comment:
I think we might also need to set the buffer_size here. Besides just memcpy, if the results returned from a read are smaller than the user buffers the size has to be updated to reflect the valid size.

Reply:
do you mean something like:

buffer_size_in_bytes  = existingBufferSize ?

also for the offset size?

Reply:
Actually I mean the opposite, `*existingBufferSize = buffer_size_in_bytes` and `*existingBufferOffsetSize = offset_buffer_size_in_bytes`. Since we need to update the sizes so tiledb/user knows exactly how much data is in the returned dataset.